src/app/services/course.service.ts

```diff
@@ -23,18 +23,24 @@ export class CourseService {
   }
 
   createItem(course: Course): Observable<Course> {
-    course.id = `${course.name}-${this.courses.length}`;
-    this.courses.push(course);
-    return of(course);
+    return this.http.post<Course>(this.courseUrl, course);
   }
 
   getItemById(id: string): Observable<Course> {
-    return this.getList().pipe(
+    const item = this.getList().pipe(
       map((courses: Course[]) => courses.find((course) => course.id === id))
     );
+    return item;
   }
 
-  search(searchPhrase: string) {
+  search(searchPhrase: Observable<string>) {
```

Comment:
`Debounce` here acts as a delay between calling the method and sending a request.
But it should prevent odd request being sent until the user stops typing for a while